exonum/src/encoding/fields.rs

```diff
@@ -239,25 +239,117 @@ impl<'a> Field<'a> for DateTime<Utc> {
     }
 
     unsafe fn read(buffer: &'a [u8], from: Offset, to: Offset) -> Self {
-        let secs = LittleEndian::read_i64(&buffer[from as usize..from as usize + 8]);
-        let nanos = LittleEndian::read_u32(&buffer[from as usize + 8..to as usize]);
+        let secs =
+            LittleEndian::read_i64(&buffer[from as usize..from as usize + mem::size_of::<i64>()]);
+        let nanos =
+            LittleEndian::read_u32(&buffer[from as usize + mem::size_of::<i64>()..to as usize]);
         Utc.timestamp(secs, nanos)
     }
 
     fn write(&self, buffer: &mut Vec<u8>, from: Offset, to: Offset) {
         let secs = self.timestamp();
         let nanos = self.timestamp_subsec_nanos();
         LittleEndian::write_i64(
-            &mut buffer[from as usize..to as usize - mem::size_of_val(&nanos)],
+            &mut buffer[from as usize..from as usize + mem::size_of::<i64>()],
             secs,
         );
         LittleEndian::write_u32(
-            &mut buffer[from as usize + mem::size_of_val(&secs)..to as usize],
+            &mut buffer[from as usize + mem::size_of::<i64>()..to as usize],
             nanos,
         );
     }
 }
 
+fn is_duration_representation_valid(secs: i64, nanos: i32) -> bool {
+    // Signs are checked to avoid multiple representations for same duration.
+    // Example: 4 s + 4e8 ns = 5 s - 6e8 ns.
+    if (secs < 0 && nanos > 0) || (secs > 0 && nanos < 0) {
+        return false;
+    }
+
+    // Absolute value of nanoseconds must less than 10 ** 9.
+    let nanos_per_sec = 1_000_000_000;
+    if nanos <= -nanos_per_sec || nanos >= nanos_per_sec {
```

Comment:
Yeah, I checked it at first. But I didn't like the following line:
>The absolute value of i32::min_value() cannot be represented as an i32, and attempting to calculate it will cause an overflow. This means that code in debug mode will trigger a panic on this case.
Since we are checking here that nanos are in acceptable range, we must think of I32_MIN value too.

`checked_abs` returns an Option, and use of it will lead to writing even more code.

Implementing an abs function that returns u32 for just one use seems overkill for me.